src/inmanta/deploy/scheduler.py

```diff
@@ -60,30 +115,38 @@ def __init__(
         self._work: work.ScheduledWork = work.ScheduledWork(
             requires=self._state.requires.requires_view(),
             provides=self._state.requires.provides_view(),
-            consumer_factory=self.start_for_agent,
+            new_agent_notify=self._start_for_agent,
         )
 
-        self._executor_manager = executor_manager
-
         # We uphold two locks to prevent concurrency conflicts between external events (e.g. new version or deploy request)
         # and the task executor background tasks.
         #
         # - lock to block scheduler state access (both model state and scheduled work) during scheduler-wide state updates
         #   (e.g. trigger deploy). A single lock suffices since all state accesses (both read and write) by the task runners are
         #   short, synchronous operations (and therefore we wouldn't gain anything by allowing multiple readers).
         self._scheduler_lock: asyncio.Lock = asyncio.Lock()
-        # - lock to serialize scheduler state updates (i.e. process new version)
-        self._update_lock: asyncio.Lock = asyncio.Lock()
-        self._environment = environment
+        # - lock to serialize updates to the scheduler's intent (version, attributes, ...), e.g. process a new version.
+        self._intent_lock: asyncio.Lock = asyncio.Lock()
 
         self._running = False
         # Agent name to worker task
         # here to prevent it from being GC-ed
         self._workers: dict[str, asyncio.Task[None]] = {}
-
-        self._code_manager = CodeManager(client)
-        self._environment = environment
-        self._client = client
+        # Set of resources for which a concrete non-stale deploy is in progress, i.e. we've committed for a given intent and
+        # that intent still reflects the latest resource intent
+        # Apart from the obvious, this differs from the agent queues' in-progress deploys in the sense that those are simply
+        # tasks that have been picked up while this set contains only those tasks for which we've already committed. For each
+        # deploy task, there is a (typically) short window of time where it's considered in progress by the agent queue, but
+        # it has not yet started on the actual deploy, i.e. it will still see updates to the resource intent.
+        self._deploying: set[ResourceIdStr] = set()
+        # Set of resources for which a concrete stale deploy is in progress, i.e. we've committed for a given intent and
+        # that intent has gone stale since
+        self._deploying_stale: set[ResourceIdStr] = set()
```

Comment:
Why do we need to keep track of this?

Reply:
1. We want to track the in-progress tasks, so that when we trigger a deploy, we know what's already running.
2. We want to make sure we only consider tasks as in-progress for the purposes of 1, if they are running for the latest intent, or they haven't committed to an intent yet.

This set is to inform the `ScheduledWork` that while there are tasks running for these resources, they should not be taken into consideration because they are known to be stale.

I have to confess (I meant to add a comment but I forgot) that I'm not completely satisfied with how this ended up. But I also don't see an alternative that I prefer. The knowledge of which tasks are running is in the agent queues mostly, but the semantics of what the tasks mean and what a stale deploy is, live in the scheduler. I tried to stick with those responsibilities as much as I could, but there are some rough edges to it that I couldn't get rid of.